subsys/bluetooth/host/settings.c

```diff
@@ -33,75 +32,122 @@
 LOG_MODULE_REGISTER(bt_settings);
 
 #if defined(CONFIG_BT_SETTINGS_USE_PRINTK)
-void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
+int bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
 			    const bt_addr_le_t *addr, const char *key)
 {
+	int err;
+
 	if (key) {
-		snprintk(path, path_size,
-			 "bt/%s/%02x%02x%02x%02x%02x%02x%u/%s", subsys,
-			 addr->a.val[5], addr->a.val[4], addr->a.val[3],
-			 addr->a.val[2], addr->a.val[1], addr->a.val[0],
-			 addr->type, key);
+		err = snprintk(path, path_size, "bt/%s/%02x%02x%02x%02x%02x%02x%u/%s", subsys,
+			       addr->a.val[5], addr->a.val[4], addr->a.val[3], addr->a.val[2],
+			       addr->a.val[1], addr->a.val[0], addr->type, key);
 	} else {
-		snprintk(path, path_size,
-			 "bt/%s/%02x%02x%02x%02x%02x%02x%u", subsys,
-			 addr->a.val[5], addr->a.val[4], addr->a.val[3],
-			 addr->a.val[2], addr->a.val[1], addr->a.val[0],
-			 addr->type);
+		err = snprintk(path, path_size, "bt/%s/%02x%02x%02x%02x%02x%02x%u", subsys,
+			       addr->a.val[5], addr->a.val[4], addr->a.val[3], addr->a.val[2],
+			       addr->a.val[1], addr->a.val[0], addr->type);
+	}
+
+	if (err < 0) {
+		return -EINVAL;
 	}
 
 	LOG_DBG("Encoded path %s", path);
+
+	return 0;
 }
-#else
-void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
-			    const bt_addr_le_t *addr, const char *key)
+
+static int bt_settings_encode_key_no_addr(char *path, size_t path_size, const char *key)
 {
-	size_t len = 3;
+	int err;
 
-	/* Skip if path_size is less than 3; strlen("bt/") */
-	if (len < path_size) {
-		/* Key format:
-		 *  "bt/<subsys>/<addr><type>/<key>", "/<key>" is optional
-		 */
-		strcpy(path, "bt/");
-		strncpy(&path[len], subsys, path_size - len);
-		len = strlen(path);
-		if (len < path_size) {
-			path[len] = '/';
-			len++;
-		}
+	err = snprintk(path, path_size, "bt/%s", key);
+	if (err < 0) {
+		return -EINVAL;
+	}
 
-		for (int8_t i = 5; i >= 0 && len < path_size; i--) {
-			len += bin2hex(&addr->a.val[i], 1, &path[len],
-				       path_size - len);
-		}
+	return 0;
+}
 
-		if (len < path_size) {
-			/* Type can be either BT_ADDR_LE_PUBLIC or
-			 * BT_ADDR_LE_RANDOM (value 0 or 1)
-			 */
-			path[len] = '0' + addr->type;
-			len++;
-		}
+#else /* !CONFIG_BT_SETTINGS_USE_PRINTK */
+int bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
+			   const bt_addr_le_t *addr, const char *key)
+{
+	/* Key format:
+	 *  "bt/<subsys>/<addr><type>/<key>", "/<key>" is optional
+	 */
+	const size_t prefix_len = sizeof("bt") - 1;
+	const size_t subsys_len = sizeof('/') + strlen(subsys);
+	const size_t addr_str_len = sizeof('/') + sizeof("554433221100t") - 1;
+	const size_t key_len = key == NULL ? 0 : sizeof('/') + strlen(key);
+	const size_t total_len = prefix_len + subsys_len + addr_str_len + key_len + sizeof('\0');
```

Comment:
Probably needs something like 

```suggestion
	const size_t prefix_len = sizeof("bt") - 1;
	const size_t subsys_len = strlen(subsys);
	const size_t addr_str_len = sizeof("554433221100t") - 1;
	const size_t key_len = key == NULL ? 0 : strlen(key);
	const size_t total_len = prefix_len + sizeof('/') + subsys_len + sizeof('/') + addr_str_len + key == NULL ? 0 : (sizeof('/') + key_len) + sizeof('\0');
```

To deal with the sonar issues. I guess we don't have any CI tests for `CONFIG_BT_SETTINGS_USE_PRINTK=n` to verify correctness of these functions?